### DISCUSSION TITLE
Qute: why use fragments instead of hx-boost?

### DISCUSSION THREAD

**Discussion by razshare** [quarkusio/quarkus#41114] [Q&A] [answered]
In a [recent Quarkus Insights](https://www.youtube.com/watch?v=f3dbxzgf6i8), Qute and Qute fragments were showcased.

I'm not very familiar with Qute itself but it looks very promising and I think I want to use it at my company given the chance.\
\
I would say it's best quality so far is that it's type safe and at the same time it looks like a simple to use DSL, which is perfect imo because I already have enough to think about in my java files, I want my templates as simple as possible.

So with simplicity in mind, my question is: why would I use a fragment when I can just use [hx-boost](https://htmx.org/attributes/hx-boost/) instead?

Not only it's easier to use, but it also promotes web standards and builds on top of them.
For example if a user disables JavaScript in their browser and you use hx-boost in your application, your application will keep working even if the user disables javaScript, because your DOM will just fallback to its defaults behavior, which is navigating between pages.

Take for example this snippet

```html
<form hx-boost="true" action="/example" method="post">
    <input name="email" type="email" placeholder="Enter email...">
    <button>Submit</button>
</form>
```

Suppose the user disables JS, the default behavior of `<form>` will kick in and... it will just keep working even with JS disabled.

Once you start introducing things like fragments you lose this ability.

Sure, boosting means you're fetching the whole page, but there's a difference between

1. Fetching & Re-rendering the whole page
2. Fetching the whole page & Re-rendering just the `<body>`

The second use case is the default case for hx-boost, meaning you're **not** reloading scripts and stylesheets in the `<head>` of the newly fetched page, or `frame`, let's call it that.
You're just re-rendering the `<body>`, which deals away with all the extra costs in terms of requests to the server.

The only extra cost is the chunks of markdown contained in your `<body>` element that did not update since the last frame.

---

If the advantage of fragments is just being a good citizen of the web and not wasting those extra bytes, then I can't argue, fair enough.\
\
But then in again, a viewer in that same Quarkus Insight [asked if fragments can be nested](https://youtu.be/f3dbxzgf6i8?t=2486).\
And that, to me, just sounds like crazy land to debug and probably to implement on your side as well.\
\
So are fragments worth it?\
\
I'm just trying to understand a bit what the direction of Qute is, if it aims for simplicity or not.

[a comment by an automated account is not shown]

**Comment by mkouba** [comment, marked as answer]
I think that both the [hx-boost](https://htmx.org/attributes/hx-boost/) and the fragments (the idea is described in [this essay](https://htmx.org/essays/template-fragments/)) have different use cases. It's not one or the other.

> But then in again, a viewer in that same Quarkus Insight [asked if fragments can be nested](https://youtu.be/f3dbxzgf6i8?t=2486).
And that, to me, just sounds like crazy land to debug and probably to implement on your side as well.

I checked our tests, and found [this one](https://github.com/quarkusio/quarkus/blob/3.11.1/independent-projects/qute/core/src/test/java/io/quarkus/qute/FragmentTest.java#L33-L44) and [another one](https://github.com/quarkusio/quarkus/blob/3.11.1/extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/typesafe/fragment/ComplexCheckedTemplateFragmentTest.java). So yes, fragments can be nested. But again, you don't need to use it. It depends on your specific use case.

> I'm just trying to understand a bit what the direction of Qute is, if it aims for simplicity or not.

It aims to as simple as possible, but not simpler :wink:. In other words, it's always a trade-off between being flexible/powerful and simple/convenient.

CC @ia3andy

**Reply by razshare** [reply]
Hello @mkouba ,
thank you for the reply, I appreciate it.

I think I understand.

I wasn't aware of the fact that the htmx documentation itself takes into account fragments as an actual implementation.

I stand corrected, thank you!

**Reply by mkouba** [reply]
:+1: You're welcome!
